keyboards/melody96/keymaps/zunger/keymap.c

```diff
@@ -0,0 +1,250 @@
+
+  // Function layer is mostly for keyboard meta-control operations. Lots of this is just from the
+  // default layout; TODO make it nicer, add Unicode mode switchers.
+	[_FUNCTION] = LAYOUT_hotswap(
+    RESET,   _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, KC_MUTE, KC_VOLD, KC_VOLU, _______, _______, _______,
+    _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______, _______, _______, _______, _______,
+    _______, RGB_TOG, _______, RGB_MOD, RGB_HUI, RGB_HUD, RGB_SAI, RGB_SAD, RGB_VAI, RGB_VAD, _______, _______, _______,          _______, _______, _______, _______,
+    BL_TOGG, _______, _______, UC_M_OS, UC_M_LN, UC_M_WI, UC_M_BS, UC_M_WC, _______, _______, _______, _______,                   _______, _______, _______, _______, _______,
+    _______,          _______, _______, BL_DEC,  BL_TOGG, BL_INC,  _______, _______, _______, _______, _______, _______,          _______, _______, _______, _______,
+    _______, _______, _______,                            _______,                            _______, _______, _______, _______, _______, _______, _______, _______, _______),
+};
+
+// Layer bitfields.
+#define GREEK_LAYER (1UL << _GREEK)
+#define SHIFTGREEK_LAYER (1UL << _SHIFTGREEK)
+#define CADET_LAYER (1UL << _CADET)
+#define SHIFTCADET_LAYER (1UL << _SHIFTCADET)
+// The layers we don't touch.
+#define LAYER_MASK ~(GREEK_LAYER|SHIFTGREEK_LAYER|CADET_LAYER|SHIFTCADET_LAYER)
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  // We track these persistent globals and manage them on our own, rather than trying to rely on
+  // get_mods or the like, because this function is called *before* that's updated!
+  static bool shift_held = false;
+  static bool alt_held = false;
+  static bool greek_held = false;
+  static bool cadet_held = false;
+
+  // These are where we remember the values of lock states.
+  static bool shift_lock = false;
+  static uint32_t layer_lock = _QWERTY;
+
+  // Process any modifier key presses.
+  if (keycode == KC_LSHIFT || keycode == KC_RSHIFT) {
+    shift_held = record->event.pressed;
+  } else if (keycode == KC_LALT || keycode == KC_RALT) {
+    alt_held = record->event.pressed;
+  } else if (keycode == KC_GREEK) {
+    greek_held = record->event.pressed;
+  } else if (keycode == KC_CADET) {
+    cadet_held = record->event.pressed;
+  }
+
+  // Now let's transform these into the "cadet request" and "greek request."
+  const bool greek_request = (greek_held && !alt_held);
+  const bool cadet_request = (cadet_held || (greek_held && alt_held));
+
+  // Now, handle the lock keys. We store next_layer_lock in a local variable so that we can
+  // determine the layer to pick right now before we update layer_lock.
+  uint32_t next_layer_lock = layer_lock;
```

Comment:
The layer mask isn't always 32 bit, and can be changed. 

If you could change to the typedef that we recently added? 

layer_state_t
```suggestion
  layer_state_t next_layer_lock = layer_lock;
```